R/bm-tpc-h.R

```diff
@@ -0,0 +1,206 @@
+#' Benchmark TPC-H queries
+#'
+#' @section Parameters:
+#' * `engine` One of `c("arrow", "duckdb", "dplyr)`
+#' * `query_num` integer, 1-22
+#' * `format` One of `c("parquet", "feather", "native")`
+#' * `scale` Scale factor to use for data generation (e.g. 0.1, 1, 10, 100)
+#'
+#' @export
+tpc_h <- Benchmark("tpc_h",
+  setup = function(engine = "arrow",
+                   query_num = 1,
+                   format = "parquet",
+                   scale = 10) {
+    # engine defaults to arrow
+    engine <- match.arg(engine, c("arrow", "duckdb", "dplyr"))
+    # input format
+    format <- match.arg(format, c("parquet", "feather", "native"))
+    # query_num defaults to 1 for now
+    stopifnot(
+      "query_num must be an int" = query_num %% 1 == 0,
+      "query_num must 1-22" = query_num >= 1 & query_num <= 22
+    )
+
+    library("dplyr")
+
+    # ensure that we have the _base_ tpc-h files (in parquet)
+    tpch_files <- ensure_source("tpch", scale = scale)
+
+    # these will be filled in later, when we have file paths available
+    input_functions <- list()
+
+    # we use this connection both to populate views/tables and get answer info
+    con <- DBI::dbConnect(duckdb::duckdb())
+
+    if (engine == "arrow") {
+      # ensure that we have the right kind of files available
+      # but for native, make sure we have a feather file, and we will read that
+      # in to memory before the benchmark (below)
+      format_to_convert <- format
+      if (format == "native") {
+        format_to_convert <- "feather"
+      }
+
+      tpch_files <- vapply(
+        tpch_files,
+        ensure_format,
+        FUN.VALUE = character(1),
+        format = format_to_convert
+      )
+
+      # specify readers for each format
+      if (format == "parquet") {
+        input_functions[["arrow"]] <- function(name) {
+          file <- tpch_files[[name]]
+          return(arrow::read_parquet(file, as_data_frame = FALSE))
+        }
+      } else if (format == "feather") {
+        input_functions[["arrow"]] <- function(name) {
+          file <- tpch_files[[name]]
+          return(arrow::read_feather(file, as_data_frame = FALSE))
+        }
+      } else if (format == "native") {
+        # native is different: read the feather file in first, and pass the table
+        tab <- list()
+        for (name in names(tpch_files)) {
```

Comment:
So we have to read in all tpch_files now? Do these queries share the same dataset, or do we get 22 datasets for 22 queries? 

Since each process is only running a single query (right?) it seems like you should be able to only read in the ones that the current query requires

Reply:
That is what this is doing, yeah. Each query uses the same dataset (well dataset of multiple tables). This query only needs one table (lineitem), so we really only need to read that in.

I can make a query to table map so that we only read in the tables that are necessary for each query.

Ultimately, I don't think this will have all that much of an impact, lineitem is the largest table  by far, so reading the others in shouldn't have a huge impact even if they aren't used.

Additionally, right now what happens is that each of the files are memory mapped by this, but aren't actually paged into memory until the first run (this is why for arrow the first run in this condition is so much longer than the others). I'm planning to turn memory mapping off so that we really are starting with a truly in-memory table (though we could parameterize that if we wanted to!)